mmf/models/krisp.py

```diff
@@ -0,0 +1,764 @@
+
+        elif self.config.graph_logit_mode == "in_graph":
+            # Logits is already computed
+            graph_logits = graph_output
+            assert self.config.graph_module.output_type == "graph_prediction"
+        elif self.config.graph_logit_mode == "logit_fc":
+            # Compute logits from single hidden layer
+            graph_logits = self.graph_logit_fc(graph_output)
+
+        # Now combine outputs
+        if self.config.output_combine == "concat":
+            # Output order should be alphabetical
+            assert self.config.graph_module.output_order == "alpha"
+
+            # Combine both logits
+            logits = torch.cat([vb_logits, graph_logits], dim=1)
+        elif self.config.output_combine == "add":
+            # Output order should be ans
+            assert self.config.graph_module.output_order == "ans"
+
+            # Set invalid inds to zero here
+            assert graph_logits.size(1) == vb_logits.size(1)
+            graph_logits[:, self.missing_ans_inds] = 0
+            logits = vb_logits + graph_logits
+
+        # Do zerobias
+        if self.config.zerobias:
+            logits -= 6.58
+
+        # For loss calculations (automatically done by MMF
+        # as per the loss defined in the config),
+        # we need to return a dict with "scores" key as logits
+        output = {"scores": logits}
+
+        # If we're in eval / analysis mode, add more to output
+        if self.config.analysis_mode:
+            output = self.graph_module.add_analysis_to_output(output)
+
+        # MMF will automatically calculate loss
+        return output
+
+
+class GraphPtrNet(nn.Module):
+    def __init__(self, hidden_size, graph_hidden_size):
+        super().__init__()
+
+        self.hidden_size = hidden_size
+        self.graph_hidden_size = graph_hidden_size
+
+        self.bl_w = nn.Linear(hidden_size, hidden_size)
+        self.graph_w = nn.Linear(graph_hidden_size, hidden_size)
+
+    def forward(self, bl_hidden, graph_hidden):
+        # Compute Eq. 4 from Iterative Answer Prediction with Pointer-Augmented Multimodal Transformers for TextVQA
+        # bl_hidden is bs x hidden_size
+        # graph_hidden is bs x graph_hidden_size
+
+        # Compute BL half
+        bl_hidden = self.bl_w(bl_hidden)
+        assert bl_hidden.dim() == 2
+        bl_hidden = bl_hidden.unsqueeze(1)
+
+        # Compute graph hidden half
+        # Assume we've already subsampled to only valid answer nodes
+        graph_hidden = self.graph_w(graph_hidden)
+
+        # Now we have bl_hidden as a bs x 1 x hid vec
+        # graph_hidden as a bs x num_nodes x hid vec
+
+        # Combine
+        scores = torch.matmul(bl_hidden, graph_hidden.transpose(-1, -2))
+
+        # Normalize
+        scores = scores / math.sqrt(self.hidden_size)
+        scores = scores.squeeze(1)
+
+        # Scores is now a bs x #nodes matrix
+        return scores
+
+
+class VisualBERTBase(BertPreTrainedModel):
```

Comment:
There is a lot of code duplication going here. Is it super-diverged that you need a custom implementation or can you inherit and get away with it?

Reply:
Yeah I know, but there's no way to fix that without rewriting VisualBERT. Need to remove the last part of VisualBERT to make mine work.

Reply:
Can you comment here specifically which parts are changed in this version of VisualBERT?